firewood/src/shale/mod.rs

```diff
@@ -149,7 +149,9 @@ impl<T: Storable> Obj<T> {
 
 impl<T: Storable> Drop for Obj<T> {
     fn drop(&mut self) {
-        self.flush_dirty()
+        if self.dirty.is_some() {
+            self.flush_dirty()
+        }
```

Comment:
Upon further review, it appears that `is_mem_mapped()` will always return false (part of the `flush_dirty` method above). That method should be removed entirely. It's a trait method (part of `Storable`), which means there's an indirection every time it's called and it's possible that the optimizer has a hard time figuring out that there isn't a single non-default implementation of the method (which just defaults to `false`).

Further, I wouldn't expect `self.dirty.take()` to take longer if `self.dirty == None`, so you shouldn't need that extra check. Did you see the 3% drop in time in `release` mode? 

I suspect that part of the slowdown here is also due to the double indirection in `StoredView::mem` (the property). Its type is `Box<dyn SendSyncDerefMut<Target = dyn CachedStore>>`. We have to go to two different vtables just to call the `CachedStore::write` trait-method. 

I wonder how difficult it would be to remove that `dyn SendSyncDerefMut` though? It's definitely a type of lifetime hack.

Anyway, enough thinking out loud. The one concrete thing that should be changed and measured is the removal of `is_mem_mapped` and see if you get a decrease in time spend. 

Please post the result in the conversation section along with the compiler flags used so that we can keep ourselves honest about compiling in release-mode.

Reply:
yeah, I'm using release mode as that's the default for cargo `flamegraph`.

Remove is_mem_mapped() is not showing perf decrease, so added it as well. But I do think `self.dirty.take()` has a cost here even for `self.dirty == None`, probably because this is a very critical path even a tiny diff is shown.